scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/policy/edf/EDFPolicyExtended.java

```diff
@@ -0,0 +1,157 @@
+import java.util.Comparator;
+import java.util.Date;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import org.ow2.proactive.scheduler.common.JobDescriptor;
+import org.ow2.proactive.scheduler.common.job.JobPriority;
+import org.ow2.proactive.scheduler.descriptor.EligibleTaskDescriptor;
+import org.ow2.proactive.scheduler.descriptor.JobDescriptorImpl;
+import org.ow2.proactive.scheduler.job.InternalJob;
+import org.ow2.proactive.scheduler.policy.ExtendedSchedulerPolicy;
+
+
+/**
+ * Early Deadline First Policy sorts jobs based on:
+ * - job priorities (from highest to lowest)
+ * - job deadline (those that have deadline overtake those without deadline)
+ *
+ * Among the job with the dealine (and having same priority) we distinguish
+ * them by the fact if some task of the job is already scheduled (i.g. job has
+ * a startingTime) or not. Those that started has a priority to the not yet started jobs.
+ * Among the jobs that are already started we sort them by startTime.
+ * If job is not started, we sort them by `effectiveDeadline - (now() + expectedTime`.
+ * If deadline is absolute then effectiveDeadline equals to deadline.
+ * If deadline is relative then effectiveDeadline is calculated as (now() + deadline)
+ *
+ * Among jobs without deadline (and the same priority), we sort them by submission date.
+ *
+ * @author The ProActive Team
+ * @since ProActive Scheduling 8.3
+ */
+public class EDFPolicyExtended extends ExtendedSchedulerPolicy {
+
+    private static final Date MAXIMUM_DATE = new Date(Long.MAX_VALUE);
+
+    @Override
+    public LinkedList<EligibleTaskDescriptor> getOrderedTasks(List<JobDescriptor> jobs) {
+        final Date now = new Date();
+
+        final Comparator<JobDescriptor> jobDescriptorComparator = (job1, job2) -> {
+
+            final InternalJob internalJob1 = ((JobDescriptorImpl) job1).getInternal();
+            final InternalJob internalJob2 = ((JobDescriptorImpl) job2).getInternal();
+            JobPriority job1Priority = internalJob1.getPriority();
+            JobPriority job2Priority = internalJob2.getPriority();
+            if (!job1Priority.equals(job2Priority)) {
+                return job2Priority.compareTo(job1Priority);
+            } else {
+                if (internalJob1.getJobDeadline().isPresent() & !internalJob2.getJobDeadline().isPresent()) {
+                    return -1;
+                } else if (!internalJob1.getJobDeadline().isPresent() & internalJob2.getJobDeadline().isPresent()) {
+                    return 1;
+                } else if (!internalJob1.getJobDeadline().isPresent() & !internalJob2.getJobDeadline().isPresent()) {
+                    return Long.compare(internalJob1.getJobInfo().getSubmittedTime(),
+                                        internalJob2.getJobInfo().getSubmittedTime());
+                } else { // if(internalJob1.getJobDeadline().isPresent() & internalJob2.getJobDeadline().isPresent()) {
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') This block of commented-out lines of code should be removed. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3ACommentedOutCodeLine)

Reply:
+1